### DISCUSSION TITLE
OTP via email for signup

### DISCUSSION THREAD

**Discussion by potpov** [supabase/supabase#4837] [Feature Requests]
Hello, 
First of all, thank you for your great work.

I would like to ask you if there is a way to use the email sign-up with the OTP verification rather than the email link. 
(as it already happens for the registration via mobile phone).

For instance I would love to have the same API entry at the first stage:

```javascript
const { user, session, error } = await supabase.auth.signUp({
  email: 'example@email.com',
  password: 'example-password',
})
```

But then send to the user a token rather than a link, and verify the token on the same page my user was visiting:

```javascript
let { session, error } = await supabase.auth.verifyOTP({
  email: 'example@email.com',
  token: '123456',
})
```

Haven't found this on the doc, I wonder if there is already a workaround or if this can be an idea for future-features. 
thank you!

**Comment by awalias** [comment]
thanks for the question! we're tracking this one here: supabase/gotrue#290

**Reply by zxqp** [reply]
https://github.com/supabase/gotrue/releases/tag/v2.5.8

is it ready to use now?

**Reply by asyncink** [reply]
@awalias, +1 for @zxqp question

**Comment by potpov** [comment]
Alrighty, 
I assumed that was more of a discussion and didn't check the issues! 
my bad, thank you for the link!

**Reply by GaryAustin1** [reply]
I can confirm per the latest blog post: https://supabase.com/blog/2022/03/28/community-day (scroll to Send OTP via email)

Email OTP is working!

Email on signup with template modified with {{ .Token}}

![image](https://user-images.githubusercontent.com/54564956/160448771-a1f0b508-3688-4c32-a24a-8cc992e5db47.png)

And call made to confirm the user.
```
const {user,session,error} = await supabase.auth.verifyOTP ({
	email:'name@email.com',
	token:'owOWaODcyzG0ItuHQ24ipA',
	type:'signup'
})
```

**Reply by asyncink** [reply]
thank you for your reply!

I made it work:

1. request otp:

```ts
const { error } = await supabase.auth.signIn({ email })
```

2. verify otp:

```ts
const { session, error: verifyError } = await supabase.auth.verifyOTP({
  email,
  token,
  type: 'magiclink'
})
```

**Reply by shahbaz-apptech** [reply]
i am facing some issues saying first singups are not allowed with otp can any body explain that to me why

**Comment by niklasgrewe** [comment]
does anyone know if it is also possible to send only four digit or six digit codes like Slack and many others do?

The current codes are rather less suitable for manual input, or is there another reason why these are cryptic?

**Reply by wd021** [reply]
agreed. would be way more user-friendly to have easier codes w/ short timespans. These hashed codes right now make it unusable for mobile.

**Reply by GaryAustin1** [reply]
This is an old post.   There is a short, default 6 digit, numeric code using {{Token}} in the email template.  You can set the length in settings.

**Reply by shihfu** [reply]
@GaryAustin1, can you point me to the docs for this particular setting? I can't seem to find this option anywhere.

**Reply by GaryAustin1** [reply]
I can't find it anymore...

This is still in the code...
       ` if config.Mailer.OtpLength == 0 || config.Mailer.OtpLength < 6 || config.Mailer.OtpLength > 10 {`
       It is still in the gotrue code....  https://github.com/supabase/gotrue/blob/425487325b3e1d7be4661e9f292232205a9a3ce0/internal/conf/configuration.go#L577

But I can't find it or any record of it being removed.  So could be I was mistaken back then or it was removed.

**Reply by othnielussher** [reply]
@shihfu 
From your project dashboard page, locate it at **Authentication** -> **Configuration** -> **Email Templates**.
Then in the message source section, you can then replace the usually default variable _**{{ .ConfirmationURL }}**_  with **_{{ .Token }}_**

<img width="1440" alt="Screenshot 2024-02-26 at 12 57 41" src="https://github.com/supabase/supabase/assets/53237532/d436d605-5899-44c7-bfbc-be995b173429">

**Reply by shihfu** [reply]
@othnielussher, I still don't see the option to change the token length.

**Comment by ddtch** [comment]
I still have same problem. all is set up, OTP should work but magic link is giving me an error like  `AuthApiError: Signups not allowed for otp`

**Reply by GaryAustin1** [reply]
You should start a new discussion or ask on Discord.  This is very old and AuthApiError does not appear anywhere here.
You should provide the code in the new post for what call is actually failing.

Is your email provider enabled in the authentication section of the dashboard?

**Comment by WEnigmaHQ** [comment]
most of our buddies worked otp verification react native but the problem #docs expalain in vague form which confused alot. 

```
const {token, setToken} = useState('');

const {data, error} = await supabase.auth.verifyOTP({
  email,
  token,
  type: 'email'
})
```

when user register any account by default user receive email from supabase auth. In case you have clicked you have link comfirmation mail [dumby link]. 

**Authentication > Configuration > Email > Signup email**

```
<h2>Confirm your signup</h2>

<p>Follow this link to confirm your user:</p>
<p> Your token will be {{ .Token }}</p>
```

change link with token.
